contracts/staking/PoSValidatorManager.sol

```diff
@@ -130,4 +140,155 @@ abstract contract PoSValidatorManager is IPoSValidatorManager, ValidatorManager
+            $._delegatorStakes[validationID][delegator].weight,
+            weight,
+            block.timestamp
+        );
+    }
+
+    function initializeEndDelegation(bytes32 validationID) external {
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+
+        // Ensure the delegator is active
+        Delegator memory delegator = $._delegatorStakes[validationID][_msgSender()];
+        require(
+            delegator.status == ValidatorStatus.Active, "ValidatorManager: Delegator not active"
+        );
+        delegator.status = ValidatorStatus.PendingRemoved;
+        delegator.endedAt = uint64(block.timestamp);
+
+        $._delegatorStakes[validationID][_msgSender()] = delegator;
+
+        Validator memory validator = _getValidator(validationID);
+        require(validator.weight >= delegator.weight, "PoSValidatorManager: Invalid weight");
+
+        // Submit the message to the Warp precompile.
+        bytes memory setValidatorWeightPayload = ValidatorMessages
+            .packSetSubnetValidatorWeightMessage(
+            validationID, _getAndIncrementNonce(validationID), validator.weight - delegator.weight
+        );
+        $._pendingEndDelegatorMessages[validationID][_msgSender()] = setValidatorWeightPayload;
+        bytes32 messageID = WARP_MESSENGER.sendWarpMessage(setValidatorWeightPayload);
+
+        emit DelegatorRemovalInitialized(validationID, messageID, _msgSender(), block.timestamp);
+    }
+
+    function resendEndDelegation(bytes32 validationID, address delegator) external {
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+        require(
+            $._pendingEndDelegatorMessages[validationID][delegator].length > 0
+                && $._delegatorStakes[validationID][delegator].status == ValidatorStatus.PendingRemoved,
+            "PoSValidatorManager: Delegator removal not pending"
+        );
+        WARP_MESSENGER.sendWarpMessage($._pendingEndDelegatorMessages[validationID][delegator]);
+    }
+
+    function completeEndDelegation(uint32 messageIndex, address delegator) external {
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+
+        // Unpack the Warp message
+        WarpMessage memory warpMessage = _getPChainWarpMessage(messageIndex);
+        (bytes32 validationID, uint64 nonce, uint64 weight) =
+            ValidatorMessages.unpackSubnetValidatorWeightUpdateMessage(warpMessage.payload);
+        require(
+            $._pendingEndDelegatorMessages[validationID][delegator].length > 0
+                && $._delegatorStakes[validationID][delegator].status == ValidatorStatus.PendingRemoved,
+            "PoSValidatorManager: Delegator removal not pending"
+        );
+        delete $._pendingEndDelegatorMessages[validationID][delegator];
+
+        // Update the validator weight and delegator status
+        Validator memory validator = _getValidator(validationID);
+        require(
```

Comment:
same here, don't think this will work for multiple delegation ends to the same validation ID